src/pip/_internal/wheel.py

```diff
@@ -772,61 +772,90 @@ def _contains_egg_info(
     return bool(_egg_info_re.search(s))
 
 
-def should_use_ephemeral_cache(
+def should_build(
     req,  # type: InstallRequirement
-    should_unpack,  # type: bool
-    cache_available,  # type: bool
+    need_wheel,  # type: bool
     check_binary_allowed,  # type: BinaryAllowedPredicate
 ):
     # type: (...) -> Optional[bool]
-    """Return whether to build an InstallRequirement object using the
-    ephemeral cache.
-
-    :param cache_available: whether a cache directory is available for the
-        should_unpack=True case.
-
-    :return: True or False to build the requirement with ephem_cache=True
-        or False, respectively; or None not to build the requirement.
-    """
+    """Return whether an InstallRequirement should be built into a wheel."""
     if req.constraint:
         # never build requirements that are merely constraints
-        return None
+        return False
     if req.is_wheel:
-        if not should_unpack:
+        if need_wheel:
             logger.info(
                 'Skipping %s, due to already being wheel.', req.name,
             )
-        return None
-    if not should_unpack:
-        # i.e. pip wheel, not pip install;
-        # return False, knowing that the caller will never cache
-        # in this case anyway, so this return merely means "build it".
-        # TODO improve this behavior
         return False
 
+    if need_wheel:
+        # i.e. pip wheel, not pip install
+        return True
+
     if req.editable or not req.source_dir:
-        return None
+        return False
 
     if not check_binary_allowed(req):
         logger.info(
             "Skipping wheel build for %s, due to binaries "
             "being disabled for it.", req.name,
         )
-        return None
+        return False
+
+    return True
+
+
+def should_cache(
+    req,  # type: InstallRequirement
+    check_binary_allowed,  # type: BinaryAllowedPredicate
+):
+    # type: (...) -> Optional[bool]
+    """
+    Return whether a built InstallRequirement can be stored in the persistent
+    wheel cache, assuming the wheel cache is available, and should_build()
+    has determined a wheel needs to be built.
+    """
+    if not should_build(
+        req, need_wheel=False, check_binary_allowed=check_binary_allowed
+    ):
+        # never cache if pip install (need_wheel=False) would not have built
+        # (editable mode, etc)
+        return False
```

Comment:
I would drop this check. I think we can assume that `should_cache` would only be called if we know we're already going to be building a wheel.

Reply:
@chrahunt thanks for reviewing! 

Actually this test is important, to make sure we will not cache the wheel built for an editable (when running `pip wheel -e ...`). It's the part where `should_build` and `should_cache` semantics of `should_use_ephemeral_cache` [overlap](https://github.com/pypa/pip/blob/master/src/pip/_internal/wheel.py#L800-L815). It's important to keep that part in `should_cache` too in anticipation of #6852.

Notice it calls `should_build` with `need_wheel=False`. This way of testing this case is introduced in the [second commit](https://github.com/pypa/pip/pull/7262/commits/c5a071a0bf8eb978459534f8f410ca71650790b8..283eb23594f9af8ee56b2d2e986e3a201a9887da) as alternative implementation to avoid repeating the same tests (for editable and binary allowed) that are already in `should_build`. The result is the same without this second commit, which was made following [this review comment](https://github.com/pypa/pip/pull/6853#discussion_r317510741) and discussion.

Reply:
I think the new logic is easier to follow in any case so I'm OK with deferring duplicating the checks until it looks like it gives an obvious improvement, maybe in a later PR.